

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>DiffMethylTools &mdash; DiffMethylTools 0.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/theme_edit.css?v=8cf43333" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=47de8214"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DiffMethylTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">v0.0.3</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DiffMethylTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Module code</a></li>
      <li class="breadcrumb-item active">DiffMethylTools</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for DiffMethylTools</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Type</span><span class="p">,</span> <span class="n">get_args</span><span class="p">,</span> <span class="n">get_origin</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">from</span> <span class="nn">types</span> <span class="kn">import</span> <span class="n">UnionType</span>
<span class="kn">from</span> <span class="nn">lib</span> <span class="kn">import</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">Analysis</span><span class="p">,</span> <span class="n">Plots</span><span class="p">,</span> <span class="n">FormatDefinition</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">mannwhitneyu</span>
<span class="kn">import</span> <span class="nn">inspect</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<div class="viewcode-block" id="DiffMethylTools">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools">[docs]</a>
<span class="k">class</span> <span class="nc">DiffMethylTools</span><span class="p">():</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pipeline</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="o">=</span> <span class="n">pipeline</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">obj</span> <span class="o">=</span> <span class="n">Analysis</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span> <span class="o">=</span> <span class="n">Plots</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">def</span> <span class="nf">__prepare_parameters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameters</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;self&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">parameters</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">parameters</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
        
        <span class="k">return</span> <span class="n">parameters</span>

    <span class="n">MERGE_TABLES_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;case_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ctr_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.merge_tables">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.merge_tables">[docs]</a>
    <span class="k">def</span> <span class="nf">merge_tables</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">case_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">ctr_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">min_cov</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">cov_percentile</span> <span class="o">=</span> <span class="mf">100.0</span><span class="p">,</span> <span class="n">min_samp_ctr</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">min_samp_case</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Merge case and control data tables.</span>

<span class="sd">        .. note::</span>
<span class="sd">            ``case_data`` and ``ctr_data`` must be from a list of samples, where each contains one of the following column formats:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;coverage&quot;, &quot;methylation_percentage&quot;]``</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;positive_methylation_count&quot;, &quot;negative_methylation_count&quot;]``</span>
<span class="sd">                - Additionally, ``&quot;position_end&quot;`` and ``&quot;strand&quot;`` columns can be used. These will be considered during the join process. </span>

<span class="sd">        :param case_data: The case data to be merged.</span>
<span class="sd">        :type case_data: InputProcessor</span>
<span class="sd">        :param ctr_data: The control data to be merged.</span>
<span class="sd">        :type ctr_data: InputProcessor</span>
<span class="sd">        :param min_cov: Minimum coverage filter, defaults to 10</span>
<span class="sd">        :type min_cov: int, optional</span>
<span class="sd">        :param cov_percentile: Maximum coverage filter (percentile of sample coverage). Ranges from 0.0-100.0, defaults to 100.0</span>
<span class="sd">        :type cov_percentile: float, optional</span>
<span class="sd">        :param min_samp_ctr: Minimum samples in control, defaults to 2</span>
<span class="sd">        :type min_samp_ctr: int, optional</span>
<span class="sd">        :param min_samp_case: Minimum samples in case, defaults to 2</span>
<span class="sd">        :type min_samp_case: int, optional</span>
<span class="sd">        :return: Merged data table</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">case_data</span> <span class="o">=</span> <span class="n">case_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">ctr_data</span> <span class="o">=</span> <span class="n">ctr_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;process&quot;</span><span class="p">)</span>
        <span class="n">case_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;process2&quot;</span><span class="p">)</span>
        <span class="n">ctr_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;parameters&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">case_data</span><span class="o">=</span><span class="n">case_data</span><span class="o">.</span><span class="n">data_container</span><span class="p">,</span> <span class="n">ctr_data</span><span class="o">=</span><span class="n">ctr_data</span><span class="o">.</span><span class="n">data_container</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;function&quot;</span><span class="p">)</span>
        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;return&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>

    
    <span class="n">WINDOW_BASED_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage*&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_case&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_ctr&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.window_based">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.window_based">[docs]</a>
    <span class="k">def</span> <span class="nf">window_based</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">statistical_test</span> <span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="n">mannwhitneyu</span><span class="p">,</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">step</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span> <span class="n">min_nbr_per_win</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">min_std</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Perform window-based analysis.</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            ``data`` must contain the following column format:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;methylation_percentage*&quot;, &quot;avg_case&quot;, &quot;avg_ctr&quot;]``</span>
<span class="sd">                - There must be a ``methylation_percentage`` column for each sample to test.</span>

<span class="sd">        .. note::</span>
<span class="sd">            ``statistical_test`` must be a function from the `scipy.stats &lt;https://docs.scipy.org/doc/scipy/reference/stats.html&gt;`_ package. Accepted functions include:</span>
<span class="sd">                - ``mannwhitneyu``</span>
<span class="sd">                - ``ttest_ind``</span>
<span class="sd">                - ``ranksums``</span>
<span class="sd">                - ``ks_2samp``</span>
<span class="sd">                - ``median_test``</span>
<span class="sd">                - ``brunnermunzel``</span>

<span class="sd">        :param data: Input data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param statistical_test: Statistical test to use, must be from `scipy.stats &lt;https://docs.scipy.org/doc/scipy/reference/stats.html&gt;`_ package, defaults to ``mannwhitneyu``</span>
<span class="sd">        :type statistical_test: Callable, optional</span>
<span class="sd">        :param window: Sample window size (in bp), defaults to 1000</span>
<span class="sd">        :type window: int, optional</span>
<span class="sd">        :param step: Window step size, defaults to 500</span>
<span class="sd">        :type step: int, optional</span>
<span class="sd">        :param min_nbr_per_win: Minimum number of sample positions per window, defaults to 5</span>
<span class="sd">        :type min_nbr_per_win: int, optional</span>
<span class="sd">        :param processes: Number of CPU processes, defaults to 12</span>
<span class="sd">        :type processes: int, optional</span>
<span class="sd">        :param min_std: Minimum standard deviation filter, defaults to 0.1</span>
<span class="sd">        :type min_std: float, optional</span>
<span class="sd">        :return: Window-based analysis results</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
            
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">window_based</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>

    
    <span class="n">POSITION_BASED_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage*&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.position_based">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.position_based">[docs]</a>
    <span class="k">def</span> <span class="nf">position_based</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;limma&quot;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">test_factor</span><span class="o">=</span><span class="s2">&quot;Group&quot;</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;eBayes&quot;</span><span class="p">,</span> <span class="n">min_std</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">fill_na</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Perform position-based analysis. Has options for using the gamma function, or the limma R package.</span>

<span class="sd">        .. note::</span>
<span class="sd">            ``data`` must contain the following column format:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;methylation_percentage*&quot;]``</span>
<span class="sd">                - There must be a ``methylation_percentage`` column for each sample to test.</span>

<span class="sd">        :param data: Input data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param method: Position-based method to use, options are ``[&quot;gamma&quot;, &quot;limma&quot;]``, defaults to &quot;limma&quot;</span>
<span class="sd">        :type method: str, optional</span>
<span class="sd">        :param features: Features .csv file used with the &quot;limma&quot; method, has indicies with the names of samples in ``data`` and columns for the features, defaults to None</span>
<span class="sd">        :type features: str, optional</span>
<span class="sd">        :param test_factor: Test factor used with the &quot;limma&quot; method, defaults to &quot;Group&quot;</span>
<span class="sd">        :type test_factor: str, optional</span>
<span class="sd">        :param processes: Number of CPU processes, defaults to 12</span>
<span class="sd">        :type processes: int, optional</span>
<span class="sd">        :param model: Limma model to use, options are ``[&quot;eBayes&quot;, &quot;treat&quot;]`` defaults to &quot;eBayes&quot;</span>
<span class="sd">        :type model: str, optional</span>
<span class="sd">        :param min_std: Minimum standard deviation filter, defaults to 0.1</span>
<span class="sd">        :type min_std: float, optional</span>
<span class="sd">        :param fill_na: Fill NA values with row group average, defaults to True</span>
<span class="sd">        :return: Position-based analysis results</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
            
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">position_based</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">position_based</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>


    <span class="n">MAP_WIN_2_POS_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;window_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;region_start&quot;</span><span class="p">,</span> <span class="s2">&quot;region_end&quot;</span><span class="p">],</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_case&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_ctr&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.map_win_2_pos">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.map_win_2_pos">[docs]</a>
    <span class="k">def</span> <span class="nf">map_win_2_pos</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">window_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">sub_window_size</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">sub_window_step</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">sub_window_min_diff</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">pipeline_window_result</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Map windows to positions.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``window_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;region_start&quot;, &quot;region_end&quot;]``</span>

<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;avg_case&quot;, &quot;avg_ctr&quot;]``</span>

<span class="sd">        :param window_data: Window data to map with. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type window_data: InputProcessor, optional</span>
<span class="sd">        :param position_data: Position data to map the windows to. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param processes: Number of CPU processes, defaults to 12</span>
<span class="sd">        :type processes: int, optional</span>
<span class="sd">        :param sub_window_size: Sub-window size for a deeper difference filtering, defaults to 100</span>
<span class="sd">        :type sub_window_size: int, optional</span>
<span class="sd">        :param sub_window_step: Sub-window step size, defaults to 100</span>
<span class="sd">        :type sub_window_step: int, optional</span>
<span class="sd">        :param sub_window_min_diff: Sub-window minimum difference, defaults to 0</span>
<span class="sd">        :type sub_window_min_diff: int, optional</span>
<span class="sd">        :param pipeline_window_result: The function results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;filters&quot;, &quot;generate_q_values&quot;, &quot;window_based&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type pipeline_window_result: str, optional</span>
<span class="sd">        :return: Mapped windows to positions</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">window_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">pipeline_window_result</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;filters&quot;</span><span class="p">,</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">,</span> <span class="s2">&quot;window_based&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for pipeline_window_result. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;filters&quot;&quot;, &quot;&quot;generate_q_values&quot;&quot;, &quot;&quot;window_based&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">window_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">window_data</span> <span class="o">=</span> <span class="n">window_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">window_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">window_data</span> <span class="o">=</span> <span class="n">window_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="c1"># TODO after adding the DMR function, add option for which data to use</span>
            <span class="k">if</span> <span class="n">pipeline_window_result</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pipeline_window_result</span> <span class="o">==</span> <span class="s2">&quot;filters&quot;</span><span class="p">:</span>
                <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pipeline_window_result</span> <span class="o">==</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">:</span>
                <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pipeline_window_result</span> <span class="o">==</span> <span class="s2">&quot;window_based&quot;</span><span class="p">:</span>
                <span class="n">window_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pipeline_window_result&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">window_data</span><span class="o">=</span><span class="n">window_data</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>


    <span class="n">GENERATE_Q_VALUES_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;p-val&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.generate_q_values">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.generate_q_values">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;fdr_bh&quot;</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate q-values (also referred to as adjusted p-values or FDR).</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``data``:</span>
<span class="sd">                - ``[&quot;p-val&quot;]``</span>

<span class="sd">        :param data: Input p-value data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param method: P-value adjustment method to use from `statsmodels.stats.multitest.multipletests &lt;https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.multipletests.html#statsmodels.stats.multitest.multipletests-parameters&gt;`_, defaults to &quot;fdr_bh&quot;.</span>
<span class="sd">        :type method: str, optional</span>
<span class="sd">        :param position_or_window: The position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        :return: Data with q-values (adjusted p-values or FDR)</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span>
                    <span class="n">position_or_window</span> <span class="o">=</span> <span class="s2">&quot;window&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">position_based</span><span class="p">]</span>
                    <span class="n">position_or_window</span> <span class="o">=</span> <span class="s2">&quot;position&quot;</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">position_based</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">]</span>
            
            
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>


    <span class="n">FILTERS_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;q-value&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.filters">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.filters">[docs]</a>
    <span class="k">def</span> <span class="nf">filters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">max_q_value</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">abs_min_diff</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter data by q-value and minimum difference.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``data``:</span>
<span class="sd">                - ``[&quot;q-value&quot;, &quot;diff&quot;]``</span>

<span class="sd">        :param data: Input data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param max_q_value: Maximum q-value filter, defaults to 0.05</span>
<span class="sd">        :type max_q_value: float, optional</span>
<span class="sd">        :param abs_min_diff: Absolute minimum difference filter, defaults to 25</span>
<span class="sd">        :type abs_min_diff: int, optional</span>
<span class="sd">        :param position_or_window: The position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        :return: Filtered input data</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
                    <span class="n">position_or_window</span> <span class="o">=</span> <span class="s2">&quot;window&quot;</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                    <span class="n">position_or_window</span> <span class="o">=</span> <span class="s2">&quot;position&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
        
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">filters</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span>

        <span class="k">return</span> <span class="n">res</span></div>


    <span class="n">GENERATE_DMR_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;significant_position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.generate_DMR">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.generate_DMR">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_DMR</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">significant_position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_pos</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">neutral_change_limit</span><span class="o">=</span><span class="mf">7.5</span><span class="p">,</span> <span class="n">neutral_perc</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">opposite_perc</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">significant_position_pipeline</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
        <span class="c1"># TODO ask chris what to write here for documentation</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate Differentially Methylated Regions (DMRs).</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``significant_position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;diff&quot;]``</span>
<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;diff&quot;]``</span>

<span class="sd">        :param significant_position_data: Significant position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type significant_position_data: InputProcessor, optional</span>
<span class="sd">        :param position_data: All position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param min_pos: Minimum positions, defaults to 3</span>
<span class="sd">        :type min_pos: int, optional</span>
<span class="sd">        :param neutral_change_limit: Neutral change limit, defaults to 7.5</span>
<span class="sd">        :type neutral_change_limit: float, optional</span>
<span class="sd">        :param neutral_perc: Neutral percentage, defaults to 30</span>
<span class="sd">        :type neutral_perc: int, optional</span>
<span class="sd">        :param opposite_perc: Opposite percentage, defaults to 10</span>
<span class="sd">        :type opposite_perc: int, optional</span>
<span class="sd">        :param significant_position_pipeline: The significant position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type significant_position_pipeline: str, optional</span>
<span class="sd">        :return: DMR results</span>
<span class="sd">        :rtype: tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">significant_position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">significant_position_pipeline</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for significant_position_pipeline. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">significant_position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">significant_position_data</span> <span class="o">=</span> <span class="n">significant_position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">significant_position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">significant_position_data</span> <span class="o">=</span> <span class="n">significant_position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="c1"># can be either auto, position or window</span>
            <span class="k">if</span> <span class="n">significant_position_pipeline</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">significant_position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">significant_position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">significant_position_pipeline</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">significant_position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">significant_position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;significant_position_pipeline&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">significant_position_data</span><span class="o">=</span><span class="n">significant_position_data</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">,</span> <span class="s2">&quot;cluster_df&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">,</span> <span class="s2">&quot;unclustered_dms_df&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">,</span> <span class="s2">&quot;clustered_dms_df&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">res</span></div>

    
    <span class="n">MAP_POSITIONS_TO_GENES_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;positions&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.map_positions_to_genes">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.map_positions_to_genes">[docs]</a>
    <span class="k">def</span> <span class="nf">map_positions_to_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">positions</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_regions</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">|</span><span class="nb">str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">],</span> <span class="n">min_pos_diff</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">bed_file</span><span class="o">=</span><span class="s2">&quot;CpG_gencodev42ccrenb_repeat_epic1v2hm450.bed&quot;</span><span class="p">,</span> <span class="n">gtf_file</span><span class="o">=</span><span class="s2">&quot;gencode.v42.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">,</span> <span class="n">pipeline_input_source</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Map positions to genes.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``positions``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;diff&quot;]``</span>

<span class="sd">        :param positions: Position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type positions: InputProcessor, optional</span>
<span class="sd">        :param gene_regions: Gene regions to map to. Options are any combination of ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``, defaults to ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``</span>
<span class="sd">        :type gene_regions: list[str] | str, optional</span>
<span class="sd">        :param min_pos_diff: Minimum position difference for mapping, defaults to 0</span>
<span class="sd">        :type min_pos_diff: int, optional</span>
<span class="sd">        :param bed_file: BED annotation file with unflexible input format, defaults to &quot;CpG_gencodev42ccrenb_repeat_epic1v2hm450.bed&quot;</span>
<span class="sd">        :type bed_file: str, optional</span>
<span class="sd">        :param gtf_file: GTF annotation file with unflexible input format, defaults to &quot;gencode.v42.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        :param pipeline_input_source: Pipeline input source for pipelining, options are [&quot;auto&quot;, &quot;map_win_2_pos&quot;, &quot;generate_q_values&quot;, &quot;filters&quot;], defaults to &quot;auto&quot;</span>
<span class="sd">        :type pipeline_input_source: str, optional</span>
<span class="sd">        :return: Mapped positions to genes in a list: ``[gene mapping dataframe, CCRE mapping dataframe]``</span>
<span class="sd">        :rtype: list[pd.DataFrame] </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">positions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">pipeline_input_source</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;map_win_2_pos&quot;</span><span class="p">,</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">,</span> <span class="s2">&quot;filters&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for pipeline_input_source. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;map_win_2_pos&quot;&quot;, &quot;&quot;generate_q_values&quot;&quot;, &quot;&quot;filters&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">positions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">positions</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;map_win_2_pos&quot;</span><span class="p">:</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">:</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
        
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pipeline_input_source&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">)</span></div>

    
    <span class="n">MAP_WINDOWS_TO_GENES_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>    
        <span class="s2">&quot;windows&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;region_start&quot;</span><span class="p">,</span> <span class="s2">&quot;region_end&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.map_windows_to_genes">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.map_windows_to_genes">[docs]</a>
    <span class="k">def</span> <span class="nf">map_windows_to_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">windows</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_regions</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">|</span><span class="nb">str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">],</span> <span class="n">min_pos_diff</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">bed_file</span><span class="o">=</span><span class="s2">&quot;CpG_gencodev42ccrenb_repeat_epic1v2hm450.bed&quot;</span><span class="p">,</span> <span class="n">gtf_file</span><span class="o">=</span><span class="s2">&quot;gencode.v42.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">,</span> <span class="n">enhd_thr</span> <span class="o">=</span> <span class="mi">500000</span><span class="p">,</span> <span class="n">enhp_thr</span> <span class="o">=</span> <span class="mi">50000</span><span class="p">,</span> <span class="n">prom_thr</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">pipeline_input_source</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Map windows to genes.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``windows``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;region_start&quot;, &quot;region_end&quot;, &quot;diff&quot;]``</span>

<span class="sd">        :param windows: Window data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type windows: InputProcessor, optional</span>
<span class="sd">        :param gene_regions: Gene regions to map to. Options are any combination of ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``, defaults to ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``</span>
<span class="sd">        :type gene_regions: list[str] | str, optional</span>
<span class="sd">        :param min_pos_diff: Minimum position difference for mapping, defaults to 0</span>
<span class="sd">        :type min_pos_diff: int, optional</span>
<span class="sd">        :param bed_file: BED annotation file with unflexible input, defaults to &quot;CpG_gencodev42ccrenb_repeat_epic1v2hm450.bed&quot;</span>
<span class="sd">        :type bed_file: str, optional</span>
<span class="sd">        :param gtf_file: GTF annotation file with unflexible input, defaults to &quot;gencode.v42.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        :param enhd_thr: Distal enhancer distance threshold for finding nearest gene to CCRE, defaults to 500000</span>
<span class="sd">        :type enhd_thr: int, optional</span>
<span class="sd">        :param enhp_thr: Proximal enhancer distance threshold for finding nearest gene to CCRE, defaults to 50000</span>
<span class="sd">        :type enhp_thr: int, optional</span>
<span class="sd">        :param prom_thr: Promoter enhancer distance threshold for finding nearest gene to CCRE, defaults to 2000</span>
<span class="sd">        :type prom_thr: int, optional</span>
<span class="sd">        :param processes: Number of CPU processes, defaults to 12</span>
<span class="sd">        :type processes: int, optional</span>
<span class="sd">        :param pipeline_input_source: Pipeline input source for pipelining, options are [&quot;auto&quot;, &quot;generate_DMR&quot;, &quot;generate_q_values&quot;, &quot;filters&quot;], defaults to &quot;auto&quot;</span>
<span class="sd">        :type pipeline_input_source: str, optional</span>
<span class="sd">        :return: Mapped windows to genes in a list: ``[gene mapping dataframe, CCRE mapping dataframe]``</span>
<span class="sd">        :rtype: list[pd.DataFrame]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">windows</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">pipeline_input_source</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;generate_DMR&quot;</span><span class="p">,</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">,</span> <span class="s2">&quot;filters&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for pipeline_input_source. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;generate_DMR&quot;&quot;, &quot;&quot;generate_q_values&quot;&quot;, &quot;&quot;filters&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">windows</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">windows</span> <span class="o">=</span> <span class="n">windows</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">windows</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">windows</span> <span class="o">=</span> <span class="n">windows</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">,</span> <span class="s2">&quot;clustered_dms_df&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
            <span class="k">elif</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;generate_DMR&quot;</span><span class="p">:</span>
                <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_DMR</span><span class="p">,</span> <span class="s2">&quot;clustered_dms_df&quot;</span><span class="p">)]</span>
            <span class="k">elif</span> <span class="n">pipeline_input_source</span> <span class="o">==</span> <span class="s2">&quot;generate_q_values&quot;</span><span class="p">:</span>
                <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">windows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
        
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pipeline_input_source&quot;</span><span class="p">)</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">)</span></div>

    
    <span class="n">VOLCANO_PLOT_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;q-value&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.volcano_plot">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.volcano_plot">[docs]</a>
    <span class="k">def</span> <span class="nf">volcano_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;volcano_plot.png&quot;</span><span class="p">,</span> <span class="n">threshold</span> <span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">line</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span> <span class="n">x_range</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">),</span> <span class="n">y_max</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a volcano plot.</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``data``:</span>
<span class="sd">                - ``[&quot;q-value&quot;, &quot;diff&quot;]``</span>

<span class="sd">        :param data: Input data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param name: Output file name, defaults to &quot;volcano_plot.png&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param threshold: Q-value threshold for horizontal line. Set to None to have no line, defaults to 0.05</span>
<span class="sd">        :type threshold: float, optional</span>
<span class="sd">        :param line: Vertical line threshold for the `abs(line)` vertical line. Set to NOne to have no lines, defaults to 15</span>
<span class="sd">        :type line: float, optional</span>
<span class="sd">        :param x_range: X-axis range, defaults to (-100, 100)</span>
<span class="sd">        :type x_range: tuple[int, int], optional</span>
<span class="sd">        :param y_max: Y-axis maximum, defaults to None</span>
<span class="sd">        :type y_max: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param x_label: X-axis label, defaults to None for a generic label</span>
<span class="sd">        :type x_label: str, optional</span>
<span class="sd">        :param y_label: Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type y_label: str, optional</span>
<span class="sd">        :param position_or_window: The position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
            
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">MANHATTAN_PLOT_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;q-value&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;region_start&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.manhattan_plot">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.manhattan_plot">[docs]</a>
    <span class="k">def</span> <span class="nf">manhattan_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;manhattan_plot.png&quot;</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a manhattan plot.</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``data`` are in one of the following column formats:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;q-value&quot;, &quot;position_start&quot;]``</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;q-value&quot;, &quot;region_start&quot;]``</span>

<span class="sd">        :param data: Input data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type data: InputProcessor, optional</span>
<span class="sd">        :param name: Output file name, defaults to &quot;manhattan_plot.png&quot;</span>
<span class="sd">        :type name: str</span>
<span class="sd">        :param threshold: Q-value threshold for horizontal line. Set to None to have no line, defaults to 0.05</span>
<span class="sd">        :type threshold: float, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param x_label: X-axis label, defaults to None for a generic label</span>
<span class="sd">        :type x_label: str, optional</span>
<span class="sd">        :param y_label: Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type y_label: str, optional</span>
<span class="sd">        :param position_or_window: The position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>        
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>


        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pipeline is in use here, due to assert and data is None</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">)]</span>
                
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">manhattan_plot</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">COVERAGE_PLOT_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;case_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ctr_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.coverage_plot">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.coverage_plot">[docs]</a>
    <span class="k">def</span> <span class="nf">coverage_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">case_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">ctr_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">name</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;coverage_plot.png&quot;</span><span class="p">,</span> <span class="n">cov_min</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cov_max</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">cov_max_percentile</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">99.5</span><span class="p">,</span> <span class="n">bins</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a coverage plot.</span>

<span class="sd">        .. note::</span>
<span class="sd">            ``case_data`` and ``ctr_data`` must be from a list of samples, where each contains one of the following column formats:</span>
<span class="sd">                - ``[&quot;coverage&quot;]``</span>
<span class="sd">                - ``[&quot;positive_methylation_count&quot;, &quot;negative_methylation_count&quot;]``</span>

<span class="sd">        :param case_data: The case data to plot.</span>
<span class="sd">        :type case_data: InputProcessor</span>
<span class="sd">        :param ctr_data: The control data to plot.</span>
<span class="sd">        :type ctr_data: InputProcessor</span>
<span class="sd">        :param name: Output file name, defaults to &quot;coverage_plot.png&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param cov_min: Minimum coverage display, defaults to 1</span>
<span class="sd">        :type cov_min: int, optional</span>
<span class="sd">        :param cov_max: Maximum coverage display, defaults to -1</span>
<span class="sd">        :type cov_max: int, optional</span>
<span class="sd">        :param cov_max_percentile: Maximum coverage percentile display. Ranges from 0.0-100.0, defaults to 99.5. Overrides ``cov_max`` if set.</span>
<span class="sd">        :type cov_max_percentile: float, optional</span>
<span class="sd">        :param bins: Number of bins, defaults to 20</span>
<span class="sd">        :type bins: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param x_label: X-axis label, defaults to None for a generic label</span>
<span class="sd">        :type x_label: str, optional</span>
<span class="sd">        :param y_label: Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type y_label: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">case_data</span> <span class="o">=</span> <span class="n">case_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">ctr_data</span> <span class="o">=</span> <span class="n">ctr_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">case_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
        <span class="n">ctr_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">case_data</span><span class="o">=</span><span class="n">case_data</span><span class="o">.</span><span class="n">data_container</span><span class="p">,</span> <span class="n">ctr_data</span><span class="o">=</span><span class="n">ctr_data</span><span class="o">.</span><span class="n">data_container</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">coverage_plot</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">GRAPH_GENE_REGIONS_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gene_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;intron_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream_diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ccre_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CCRE&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE_diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.graph_gene_regions">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.graph_gene_regions">[docs]</a>
    <span class="k">def</span> <span class="nf">graph_gene_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">ccre_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span><span class="s2">&quot;gene_regions.png&quot;</span><span class="p">,</span> <span class="n">gene_regions</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">|</span><span class="nb">str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">],</span> <span class="n">intron_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">exon_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">upstream_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">CCRE_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">intron_y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">exon_y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">upstream_y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">CCRE_y_label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a graph of gene regions.</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``gene_data``:</span>
<span class="sd">                - If ``&quot;intron&quot;`` is in ``gene_regions``: ``[&quot;intron&quot;, &quot;intron_diff&quot;]``</span>
<span class="sd">                - If ``&quot;exon&quot;`` is in ``gene_regions``: ``[&quot;exon&quot;, &quot;exon_diff&quot;]``</span>
<span class="sd">                - If ``&quot;upstream&quot;`` is in ``gene_regions``: ``[&quot;upstream&quot;, &quot;upstream_diff&quot;]``</span>
<span class="sd">            Required columns for ``ccre_data``:</span>
<span class="sd">                - If ``&quot;CCRE&quot;`` is in ``gene_regions``: ``[&quot;CCRE&quot;, &quot;CCRE_diff&quot;]`` </span>

<span class="sd">        :param gene_data: Gene data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type gene_data: InputProcessor, optional</span>
<span class="sd">        :param ccre_data: CCRE data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type ccre_data: InputProcessor, optional</span>
<span class="sd">        :param name: Output file name, defaults to &quot;gene_regions.png&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param gene_regions: Gene regions to map to. Options are any combination of ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``, defaults to ``[&quot;intron&quot;, &quot;exon&quot;, &quot;upstream&quot;, &quot;CCRE&quot;]``</span>
<span class="sd">        :type gene_regions: list[str] | str, optional</span>
<span class="sd">        :param intron_cutoff: Intron count vertical display cutoff, defaults to -1 (for no cutoff)</span>
<span class="sd">        :type intron_cutoff: int, optional</span>
<span class="sd">        :param exon_cutoff: Exon count vertical display cutoff, defaults to -1 (for no cutoff)</span>
<span class="sd">        :type exon_cutoff: int, optional</span>
<span class="sd">        :param upstream_cutoff: Upstream count vertical display cutoff, defaults to -1 (for no cutoff)</span>
<span class="sd">        :type upstream_cutoff: int, optional</span>
<span class="sd">        :param CCRE_cutoff: CCRE count vertical display cutoff, defaults to -1 (for no cutoff)</span>
<span class="sd">        :type CCRE_cutoff: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param x_label: X-axis label, defaults to None for a generic label</span>
<span class="sd">        :type x_label: str, optional</span>
<span class="sd">        :param intron_y_label: Intron Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type intron_y_label: str, optional</span>
<span class="sd">        :param exon_y_label: Exon Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type exon_y_label: str, optional</span>
<span class="sd">        :param upstream_y_label: Upstream Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type upstream_y_label: str, optional</span>
<span class="sd">        :param CCRE_y_label: CCRE Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type CCRE_y_label: str, optional</span>
<span class="sd">        :param position_or_window: The position-based or window-based results to use as input if DiffMethylTools is pipelined and no data is provided. Options are ``[&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;]``, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ccre_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">gene_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">ccre_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ccre_data</span> <span class="o">=</span> <span class="n">ccre_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">ccre_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">ccre_data</span> <span class="o">=</span> <span class="n">ccre_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">ccre_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">ccre_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">ccre_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ccre_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">gene_data</span><span class="o">=</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">ccre_data</span><span class="o">=</span><span class="n">ccre_data</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">graph_gene_regions</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">GRAPH_UPSTREAM_GENE_METHYLATION_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;gene_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;intron_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream_diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.graph_upstream_gene_methylation">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.graph_upstream_gene_methylation">[docs]</a>
    <span class="k">def</span> <span class="nf">graph_upstream_gene_methylation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_region</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="n">png_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;upstream_methylation.png&quot;</span><span class="p">,</span> <span class="n">csv_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;upstream_methylation.csv&quot;</span><span class="p">,</span> <span class="n">csv</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">left_distance</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="n">right_distance</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">hypermethylated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">gene_hypermethylated_min</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">window_hypermethylated_min</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">min_hypermethylated_windows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">hypomethylated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">gene_hypomethylated_max</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="n">window_hypomethylated_max</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="n">min_hypomethylated_windows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">position_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">clamp_positive</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span> <span class="n">clamp_negative</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">50</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span> <span class="s2">&quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a graph of upstream gene methylation.</span>
<span class="sd">        </span>
<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;diff&quot;]``</span>
<span class="sd">            Required columns for ``gene_data``:</span>
<span class="sd">                - If ``&quot;intron&quot;`` is in ``gene_regions``: ``[&quot;intron&quot;, &quot;intron_diff&quot;]``</span>
<span class="sd">                - If ``&quot;exon&quot;`` is in ``gene_regions``: ``[&quot;exon&quot;, &quot;exon_diff&quot;]``</span>
<span class="sd">                - If ``&quot;upstream&quot;`` is in ``gene_regions``: ``[&quot;upstream&quot;, &quot;upstream_diff&quot;]``</span>

<span class="sd">        :param position_data: Position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param gene_data: Gene data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type gene_data: InputProcessor, optional</span>
<span class="sd">        :param gene_region: Gene region, defaults to &quot;upstream&quot;</span>
<span class="sd">        :type gene_region: str, optional</span>
<span class="sd">        :param png_name: Output PNG file name, defaults to &quot;upstream_methylation.png&quot;</span>
<span class="sd">        :type png_name: str, optional</span>
<span class="sd">        :param csv_name: Output CSV file name, defaults to &quot;upstream_methylation.csv&quot;</span>
<span class="sd">        :type csv_name: str, optional</span>
<span class="sd">        :param csv: Input CSV file, defaults to None</span>
<span class="sd">        :type csv: str, optional</span>
<span class="sd">        :param left_distance: Left distance, defaults to 5000</span>
<span class="sd">        :type left_distance: int, optional</span>
<span class="sd">        :param right_distance: Right distance, defaults to 5000</span>
<span class="sd">        :type right_distance: int, optional</span>
<span class="sd">        :param window_size: Window size, defaults to 100</span>
<span class="sd">        :type window_size: int, optional</span>
<span class="sd">        :param hypermethylated: Include hypermethylated regions, defaults to True</span>
<span class="sd">        :type hypermethylated: bool, optional</span>
<span class="sd">        :param gene_hypermethylated_min: Minimum hypermethylation for genes, defaults to 20</span>
<span class="sd">        :type gene_hypermethylated_min: int, optional</span>
<span class="sd">        :param window_hypermethylated_min: Minimum hypermethylation for windows, defaults to 5</span>
<span class="sd">        :type window_hypermethylated_min: int, optional</span>
<span class="sd">        :param min_hypermethylated_windows: Minimum hypermethylated windows, defaults to 5</span>
<span class="sd">        :type min_hypermethylated_windows: int, optional</span>
<span class="sd">        :param hypomethylated: Include hypomethylated regions, defaults to True</span>
<span class="sd">        :type hypomethylated: bool, optional</span>
<span class="sd">        :param gene_hypomethylated_max: Maximum hypomethylation for genes, defaults to -20</span>
<span class="sd">        :type gene_hypomethylated_max: int, optional</span>
<span class="sd">        :param window_hypomethylated_max: Maximum hypomethylation for windows, defaults to -5</span>
<span class="sd">        :type window_hypomethylated_max: int, optional</span>
<span class="sd">        :param min_hypomethylated_windows: Minimum hypomethylated windows, defaults to 5</span>
<span class="sd">        :type min_hypomethylated_windows: int, optional</span>
<span class="sd">        :param position_count: Position count, defaults to 5</span>
<span class="sd">        :type position_count: int, optional</span>
<span class="sd">        :param clamp_positive: Positive clamp value, defaults to 50</span>
<span class="sd">        :type clamp_positive: int, optional</span>
<span class="sd">        :param clamp_negative: Negative clamp value, defaults to -50</span>
<span class="sd">        :type clamp_negative: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param gtf_file: GTF file, defaults to &quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        :param position_or_window: Position or window, options are [&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;], defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">csv</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">gene_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">elif</span> <span class="n">csv</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">elif</span> <span class="n">csv</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">gene_data</span><span class="o">=</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">graph_upstream_gene_methylation</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">GRAPH_AVERAGE_UPSTREAM_GENE_METHYLATION_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;gene_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;intron_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream_diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.graph_average_upstream_gene_methylation">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.graph_average_upstream_gene_methylation">[docs]</a>
    <span class="k">def</span> <span class="nf">graph_average_upstream_gene_methylation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_region</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="n">png_name</span><span class="o">=</span><span class="s2">&quot;average_upstream_methylation.png&quot;</span><span class="p">,</span> <span class="n">csv_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;average_upstream_methylation.csv&quot;</span><span class="p">,</span> <span class="n">csv</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">left_distance</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="n">right_distance</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">hypermethylated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">gene_hypermethylated_min</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">window_hypermethylated_min</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">min_hypermethylated_windows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">hypomethylated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">gene_hypomethylated_max</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="n">window_hypomethylated_max</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="n">min_hypomethylated_windows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">position_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">clamp_positive</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span> <span class="n">clamp_negative</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">50</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span> <span class="s2">&quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">,</span> <span class="n">position_or_window</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>   
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a graph of average upstream gene methylation.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;diff&quot;]``</span>
<span class="sd">            Required columns for ``gene_data``:</span>
<span class="sd">                - If ``&quot;intron&quot;`` is in ``gene_regions``: ``[&quot;intron&quot;, &quot;intron_diff&quot;]``</span>
<span class="sd">                - If ``&quot;exon&quot;`` is in ``gene_regions``: ``[&quot;exon&quot;, &quot;exon_diff&quot;]``</span>
<span class="sd">                - If ``&quot;upstream&quot;`` is in ``gene_regions``: ``[&quot;upstream&quot;, &quot;upstream_diff&quot;]``</span>

<span class="sd">        :param position_data: Position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param gene_data: Gene data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type gene_data: InputProcessor, optional</span>
<span class="sd">        :param gene_region: Gene region, defaults to &quot;upstream&quot;</span>
<span class="sd">        :type gene_region: str, optional</span>
<span class="sd">        :param png_name: Output PNG file name, defaults to &quot;upstream_methylation.png&quot;</span>
<span class="sd">        :type png_name: str, optional</span>
<span class="sd">        :param csv_name: Output CSV file name, defaults to &quot;upstream_methylation.csv&quot;</span>
<span class="sd">        :type csv_name: str, optional</span>
<span class="sd">        :param csv: Input CSV file, defaults to None</span>
<span class="sd">        :type csv: str, optional</span>
<span class="sd">        :param left_distance: Left distance, defaults to 5000</span>
<span class="sd">        :type left_distance: int, optional</span>
<span class="sd">        :param right_distance: Right distance, defaults to 5000</span>
<span class="sd">        :type right_distance: int, optional</span>
<span class="sd">        :param window_size: Window size, defaults to 100</span>
<span class="sd">        :type window_size: int, optional</span>
<span class="sd">        :param hypermethylated: Include hypermethylated regions, defaults to True</span>
<span class="sd">        :type hypermethylated: bool, optional</span>
<span class="sd">        :param gene_hypermethylated_min: Minimum hypermethylation for genes, defaults to 20</span>
<span class="sd">        :type gene_hypermethylated_min: int, optional</span>
<span class="sd">        :param window_hypermethylated_min: Minimum hypermethylation for windows, defaults to 5</span>
<span class="sd">        :type window_hypermethylated_min: int, optional</span>
<span class="sd">        :param min_hypermethylated_windows: Minimum hypermethylated windows, defaults to 5</span>
<span class="sd">        :type min_hypermethylated_windows: int, optional</span>
<span class="sd">        :param hypomethylated: Include hypomethylated regions, defaults to True</span>
<span class="sd">        :type hypomethylated: bool, optional</span>
<span class="sd">        :param gene_hypomethylated_max: Maximum hypomethylation for genes, defaults to -20</span>
<span class="sd">        :type gene_hypomethylated_max: int, optional</span>
<span class="sd">        :param window_hypomethylated_max: Maximum hypomethylation for windows, defaults to -5</span>
<span class="sd">        :type window_hypomethylated_max: int, optional</span>
<span class="sd">        :param min_hypomethylated_windows: Minimum hypomethylated windows, defaults to 5</span>
<span class="sd">        :type min_hypomethylated_windows: int, optional</span>
<span class="sd">        :param position_count: Position count, defaults to 5</span>
<span class="sd">        :type position_count: int, optional</span>
<span class="sd">        :param clamp_positive: Positive clamp value, defaults to 50</span>
<span class="sd">        :type clamp_positive: int, optional</span>
<span class="sd">        :param clamp_negative: Negative clamp value, defaults to -50</span>
<span class="sd">        :type clamp_negative: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param y_label: Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type y_label: str, optional</span>
<span class="sd">        :param gtf_file: GTF file, defaults to &quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        :param position_or_window: Position or window, defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">csv</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>

        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">gene_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">elif</span> <span class="n">csv</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">elif</span> <span class="n">csv</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">gene_data</span><span class="o">=</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">graph_average_upstream_gene_methylation</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">GRAPH_UPSTREAM_UCSC_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage*&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.graph_upstream_UCSC">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.graph_upstream_UCSC">[docs]</a>
    <span class="k">def</span> <span class="nf">graph_upstream_UCSC</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="s2">&quot;UCSC_graph.bedGraph&quot;</span><span class="p">,</span> <span class="n">before_tss</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a UCSC graph of upstream gene methylation.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;methylation_percentage*&quot;]``</span>
<span class="sd">                - There must be a ``methylation_percentage`` column for each sample to plot.</span>

<span class="sd">        :param gene_name: Gene name</span>
<span class="sd">        :type gene_name: str</span>
<span class="sd">        :param position_data: Position data</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param name: Output BEDGraph file name, defaults to &quot;UCSC_graph.bedGraph&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param before_tss: Distance before transcrption start site (TSS), defaults to 5000</span>
<span class="sd">        :type before_tss: int, optional</span>
<span class="sd">        :param gtf_file: GTF file, defaults to &quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">graph_upstream_UCSC</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">GRAPH_FULL_GENE_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;position_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage*&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.graph_full_gene">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.graph_full_gene">[docs]</a>
    <span class="k">def</span> <span class="nf">graph_full_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">position_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;gene_methylation_graph.png&quot;</span><span class="p">,</span> <span class="n">before_tss</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">after_tss</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">bin_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span> <span class="n">start_marker</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">end_marker</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">deviation_display</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">legend_size</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">12</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">case_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Case&quot;</span><span class="p">,</span> <span class="n">ctr_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Control&quot;</span><span class="p">,</span>  <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a graph of full gene methylation.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``position_data``:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;methylation_percentage*&quot;]``</span>
<span class="sd">                - There must be a ``methylation_percentage`` column for each sample to plot.</span>

<span class="sd">        :param gene_name: The name of the gene to graph.</span>
<span class="sd">        :type gene_name: str</span>
<span class="sd">        :param position_data: Position data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type position_data: InputProcessor, optional</span>
<span class="sd">        :param name: Output PNG file name, defaults to &quot;gene_methylation_graph.png&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param before_tss: Distance before transcription start site, defaults to 0</span>
<span class="sd">        :type before_tss: int, optional</span>
<span class="sd">        :param after_tss: Distance after transcription start site, defaults to None</span>
<span class="sd">        :type after_tss: int, optional</span>
<span class="sd">        :param bin_size: Bin size, defaults to 500</span>
<span class="sd">        :type bin_size: int, optional</span>
<span class="sd">        :param start_marker: Include start marker, defaults to True</span>
<span class="sd">        :type start_marker: bool, optional</span>
<span class="sd">        :param end_marker: Include end marker, defaults to True</span>
<span class="sd">        :type end_marker: bool, optional</span>
<span class="sd">        :param deviation_display: Display standard deviation regions, defaults to True</span>
<span class="sd">        :type deviation_display: bool, optional</span>
<span class="sd">        :param legend_size: Legend size, defaults to 12</span>
<span class="sd">        :type legend_size: int, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param x_label: X-axis label, defaults to None for a generic label</span>
<span class="sd">        :type x_label: str, optional</span>
<span class="sd">        :param y_label: Y-axis label, defaults to None for a generic label</span>
<span class="sd">        :type y_label: str, optional</span>
<span class="sd">        :param case_name: Case name in the legend, defaults to &quot;Case&quot;</span>
<span class="sd">        :type case_name: str, optional</span>
<span class="sd">        :param ctr_name: Control name in the legend, defaults to &quot;Control&quot;</span>
<span class="sd">        :type ctr_name: str, optional</span>
<span class="sd">        :param gtf_file: GTF file, defaults to &quot;gencode.v41.chr_patch_hapl_scaff.annotation.gtf&quot;</span>
<span class="sd">        :type gtf_file: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span> 
        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">position_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">position_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="n">position_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">position_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">]</span>
        
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">position_data</span><span class="o">=</span><span class="n">position_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">graph_full_gene</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">PIE_CHART_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gene_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;intron_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream_diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ccre_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CCRE&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE_diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.pie_chart">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.pie_chart">[docs]</a>
    <span class="k">def</span> <span class="nf">pie_chart</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">CCRE_data</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">regions</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">|</span><span class="nb">str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">],</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;pie_chart.png&quot;</span><span class="p">,</span> <span class="n">hypermethylated</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">hypomethylated</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">hypermehylated_min</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">hypomethylated_max</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="n">hypermethylated_title</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">hypomethylated_title</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">position_or_window</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a pie chart of gene regions.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Required columns for ``gene_data``:</span>
<span class="sd">                - If ``&quot;intron&quot;`` is in ``gene_regions``: ``[&quot;intron&quot;]``</span>
<span class="sd">                - If ``&quot;exon&quot;`` is in ``gene_regions``: ``[&quot;exon&quot;]``</span>
<span class="sd">                - If ``&quot;upstream&quot;`` is in ``gene_regions``: ``[&quot;upstream&quot;]``</span>
<span class="sd">            Required columns for ``CCRE_data``:</span>
<span class="sd">                - If ``&quot;CCRE&quot;`` is in ``gene_regions``: ``[&quot;CCRE&quot;]``</span>

<span class="sd">        :param gene_data: Gene data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type gene_data: InputProcessor, optional</span>
<span class="sd">        :param CCRE_data: CCRE data. Not necessary if the pipeline is in use, defaults to None</span>
<span class="sd">        :type CCRE_data: InputProcessor, optional</span>
<span class="sd">        :param regions: Gene regions, defaults to [&quot;intron&quot;, &quot;exon&quot;, &quot;CCRE&quot;, &quot;upstream&quot;]</span>
<span class="sd">        :type regions: list[str] | str, optional</span>
<span class="sd">        :param name: Output PNG file name, defaults to &quot;pie_chart.png&quot;</span>
<span class="sd">        :type name: str, optional</span>
<span class="sd">        :param hypermethylated: Include hypermethylated regions, defaults to True</span>
<span class="sd">        :type hypermethylated: bool, optional</span>
<span class="sd">        :param hypomethylated: Include hypomethylated regions, defaults to True</span>
<span class="sd">        :type hypomethylated: bool, optional</span>
<span class="sd">        :param hypermehylated_min: Minimum hypermethylation, defaults to 20</span>
<span class="sd">        :type hypermehylated_min: float, optional</span>
<span class="sd">        :param hypomethylated_max: Maximum hypomethylation, defaults to -20</span>
<span class="sd">        :type hypomethylated_max: float, optional</span>
<span class="sd">        :param hypermethylated_title: Hypermethylated title, defaults to None for a generic title</span>
<span class="sd">        :type hypermethylated_title: str, optional</span>
<span class="sd">        :param hypomethylated_title: Hypomethylated title, defaults to None for a generic title</span>
<span class="sd">        :type hypomethylated_title: str, optional</span>
<span class="sd">        :param title: Plot title, defaults to None for a generic title</span>
<span class="sd">        :type title: str, optional</span>
<span class="sd">        :param position_or_window: Position or window, options are [&quot;auto&quot;, &quot;position&quot;, &quot;window&quot;], defaults to &quot;auto&quot;</span>
<span class="sd">        :type position_or_window: str, optional</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">CCRE_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span> <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span><span class="p">),</span> <span class="s2">&quot;If the pipeline isn&#39;t in use, data must be provided.&quot;</span>
        <span class="k">assert</span> <span class="n">position_or_window</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;window&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid parameter for position_or_window. Options are: [&quot;&quot;auto&quot;&quot;, &quot;&quot;position&quot;&quot;, &quot;&quot;window&quot;&quot;]&quot;</span>


        <span class="n">parameters</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">gene_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">gene_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">CCRE_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">CCRE_data</span> <span class="o">=</span> <span class="n">CCRE_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">CCRE_data</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
            <span class="n">CCRE_data</span> <span class="o">=</span> <span class="n">CCRE_data</span><span class="o">.</span><span class="n">data_container</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">CCRE_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">))</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">CCRE_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># TODO invalid</span>
                    <span class="k">pass</span>
            <span class="k">elif</span> <span class="n">position_or_window</span> <span class="o">==</span> <span class="s2">&quot;position&quot;</span><span class="p">:</span>
                <span class="n">CCRE_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_positions_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">CCRE_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">saved_results</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">map_windows_to_genes</span><span class="p">,</span> <span class="s2">&quot;CCRE&quot;</span><span class="p">)]</span>
        
        <span class="n">parameters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;position_or_window&quot;</span><span class="p">)</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__prepare_parameters</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">gene_data</span><span class="o">=</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">CCRE_data</span><span class="o">=</span><span class="n">CCRE_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">pie_chart</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">)</span></div>


    <span class="n">ALL_ANALYSIS_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;case_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ctr_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_methylation_count&quot;</span><span class="p">,</span> <span class="s2">&quot;negative_methylation_count&quot;</span><span class="p">]</span>
    <span class="p">}</span>

<div class="viewcode-block" id="DiffMethylTools.all_analysis">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.all_analysis">[docs]</a>
    <span class="k">def</span> <span class="nf">all_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">case_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">ctr_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">window_based</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">min_cov</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">cov_percentile</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">min_samp_ctr</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">min_samp_case</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Run all analysis methods.</span>

<span class="sd">        .. note::</span>
<span class="sd">            ``case_data`` and ``ctr_data`` must be from a list of samples, where each contains one of the following column formats:</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;coverage&quot;, &quot;methylation_percentage&quot;]``</span>
<span class="sd">                - ``[&quot;chromosome&quot;, &quot;position_start&quot;, &quot;positive_methylation_count&quot;, &quot;negative_methylation_count&quot;]``</span>

<span class="sd">        .. note::</span>
<span class="sd">            if ``window_based`` is True, the following methods will be run:</span>
<span class="sd">                - ``merge_tables``</span>
<span class="sd">                - ``window_based``</span>
<span class="sd">                - ``generate_q_values``</span>
<span class="sd">                - ``filters``</span>
<span class="sd">                - ``map_win_2_pos``</span>
<span class="sd">            if ``window_based`` is False, the following methods will be run:</span>
<span class="sd">                - ``merge_tables``</span>
<span class="sd">                - ``position_based``</span>
<span class="sd">                - ``generate_q_values``</span>
<span class="sd">                - ``filters``</span>

<span class="sd">        :param case_data: Case data (list of files)</span>
<span class="sd">        :type case_data: InputProcessor</span>
<span class="sd">        :param ctr_data: Control data (list of files)</span>
<span class="sd">        :type ctr_data: InputProcessor</span>
<span class="sd">        :param window_based: Window-based analysis, defaults to True</span>
<span class="sd">        :type window_based: bool, optional</span>
<span class="sd">        :param prefix: Output file directory, defaults to &quot;.&quot;</span>
<span class="sd">        :type prefix: str, optional</span>
<span class="sd">        :param min_cov: Minimum coverage, defaults to 10</span>
<span class="sd">        :type min_cov: int, optional</span>
<span class="sd">        :param cov_percentile: Coverage percentile, defaults to 1.0</span>
<span class="sd">        :type cov_percentile: float, optional</span>
<span class="sd">        :param min_samp_ctr: Minimum control samples, defaults to 2</span>
<span class="sd">        :type min_samp_ctr: int, optional</span>
<span class="sd">        :param min_samp_case: Minimum case samples, defaults to 2</span>
<span class="sd">        :type min_samp_case: int, optional</span>
<span class="sd">        :return: Final significant positions</span>
<span class="sd">        :rtype: pd.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># True may take a lot of memory</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;merging&quot;</span><span class="p">)</span>
        <span class="n">merged</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merge_tables</span><span class="p">(</span><span class="n">case_data</span><span class="p">,</span> <span class="n">ctr_data</span><span class="p">,</span> <span class="n">min_cov</span><span class="o">=</span><span class="n">min_cov</span><span class="p">,</span> <span class="n">cov_percentile</span><span class="o">=</span><span class="n">cov_percentile</span><span class="p">,</span> <span class="n">min_samp_ctr</span><span class="o">=</span><span class="n">min_samp_ctr</span><span class="p">,</span> <span class="n">min_samp_case</span><span class="o">=</span><span class="n">min_samp_case</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">case_data</span><span class="p">,</span> <span class="n">ctr_data</span>
        <span class="n">merged</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/merged.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">window_based</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;windows&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">window_based</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">merged</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/windows.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;p-val&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">res</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/q_values.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">res</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/significant_windows.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;mapping&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">map_win_2_pos</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">res</span><span class="p">),</span> <span class="n">InputProcessor</span><span class="p">(</span><span class="n">merged</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/mapped_positions.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">position_based</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">merged</span><span class="p">))</span>
            <span class="k">del</span> <span class="n">merged</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/positions.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_q_values</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">res</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/q_values.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">(</span><span class="n">InputProcessor</span><span class="p">(</span><span class="n">res</span><span class="p">))</span>
            <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">/significant_positions.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">res</span></div>

    
    <span class="n">ALL_PLOTS_REQUIRED_COLUMNS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position_start&quot;</span><span class="p">,</span> <span class="s2">&quot;region_start&quot;</span><span class="p">,</span> <span class="s2">&quot;q-value&quot;</span><span class="p">,</span> <span class="s2">&quot;diff&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation_percentage*&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage&quot;</span><span class="p">],</span>
        <span class="s2">&quot;gene_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="s2">&quot;intron_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="s2">&quot;upstream_diff&quot;</span><span class="p">],</span>
        <span class="s2">&quot;ccre_data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CCRE&quot;</span><span class="p">,</span> <span class="s2">&quot;CCRE_diff&quot;</span><span class="p">]</span>
    <span class="p">}</span>
<div class="viewcode-block" id="DiffMethylTools.all_plots">
<a class="viewcode-back" href="../DiffMethylTools.html#DiffMethylTools.DiffMethylTools.all_plots">[docs]</a>
    <span class="k">def</span> <span class="nf">all_plots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">,</span> <span class="n">ccre_data</span><span class="p">:</span> <span class="n">InputProcessor</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Run all plot methods.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pipeline</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manhattan_plot</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">graph_gene_regions</span><span class="p">(</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">ccre_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">graph_upstream_gene_methylation</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">graph_average_upstream_gene_methylation</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pie_chart</span><span class="p">(</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">ccre_data</span><span class="p">)</span></div>
</div>

    



<span class="k">def</span> <span class="nf">parse_arguments</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;CLI for DiffMethylTools&quot;</span><span class="p">)</span>
    <span class="n">subparsers</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_subparsers</span><span class="p">(</span><span class="n">dest</span><span class="o">=</span><span class="s2">&quot;command&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Available commands&quot;</span><span class="p">)</span>

    <span class="n">obj</span> <span class="o">=</span> <span class="n">DiffMethylTools</span><span class="p">(</span><span class="n">pipeline</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Dynamically add subcommands for each method in DiffMethylTools</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">inspect</span><span class="o">.</span><span class="n">getmembers</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">predicate</span><span class="o">=</span><span class="n">inspect</span><span class="o">.</span><span class="n">ismethod</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">):</span>  <span class="c1"># Skip private methods</span>
            <span class="k">continue</span>

        <span class="n">names_argument_added</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">method_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Run the </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> method&quot;</span><span class="p">)</span>
        <span class="n">sig</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="n">method</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">param_name</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">sig</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">param_name</span> <span class="o">==</span> <span class="s2">&quot;self&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">arg_type</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">annotation</span> <span class="k">if</span> <span class="n">param</span><span class="o">.</span><span class="n">annotation</span> <span class="o">!=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="nb">str</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">arg_type</span><span class="p">)</span> <span class="o">==</span> <span class="n">UnionType</span><span class="p">:</span>
                <span class="n">arg_type</span> <span class="o">=</span> <span class="n">Union</span>

            <span class="k">if</span> <span class="n">arg_type</span> <span class="o">!=</span> <span class="n">InputProcessor</span> <span class="ow">and</span> <span class="n">arg_type</span> <span class="o">!=</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]:</span>
                <span class="n">default</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">default</span> <span class="k">if</span> <span class="n">param</span><span class="o">.</span><span class="n">default</span> <span class="o">!=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="k">if</span> <span class="n">param</span><span class="o">.</span><span class="n">kind</span> <span class="o">!=</span> <span class="n">param</span><span class="o">.</span><span class="n">POSITIONAL_ONLY</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">get_origin</span><span class="p">(</span><span class="n">arg_type</span><span class="p">)</span> <span class="ow">is</span> <span class="n">Optional</span><span class="p">:</span>
                        <span class="n">arg_type</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">__args__</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">arg_type</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">default</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;(default: </span><span class="si">{</span><span class="n">default</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">arg_type</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(required)&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_file&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(required, can be a list of files)&quot;</span><span class="p">)</span>
                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_has_header&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(use this flag if there is a header.)&quot;</span><span class="p">)</span>
                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_separator&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(default: ,)&quot;</span><span class="p">)</span>
                <span class="n">required_columns</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2">_REQUIRED_COLUMNS&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">data_name</span><span class="p">,</span> <span class="n">columns</span> <span class="ow">in</span> <span class="n">required_columns</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">param_name</span> <span class="o">==</span> <span class="n">data_name</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
                            <span class="k">if</span> <span class="s2">&quot;methylation_percentage&quot;</span> <span class="ow">in</span> <span class="n">column</span> <span class="ow">or</span> <span class="s2">&quot;methylation_count&quot;</span> <span class="ow">in</span> <span class="n">column</span> <span class="ow">or</span> <span class="s2">&quot;coverage&quot;</span> <span class="ow">in</span> <span class="n">column</span><span class="p">:</span>
                                <span class="c1"># remove * from column if exists</span>
                                <span class="n">column</span> <span class="o">=</span> <span class="n">column</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">column</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;*&quot;</span> <span class="k">else</span> <span class="n">column</span>
                                <span class="c1"># allow list input</span>
                                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">_column_index&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(default: None, assumes all column are named as expected)&quot;</span><span class="p">)</span>
                                <span class="k">if</span> <span class="n">param_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">names_argument_added</span><span class="p">:</span>
                                    <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_column_names&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(default: None, assumes all column are named as expected)&quot;</span><span class="p">)</span>
                                    <span class="k">if</span> <span class="s2">&quot;methylation_count&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">column</span> <span class="ow">and</span> <span class="s2">&quot;ctr&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">param_name</span> <span class="ow">and</span> <span class="s2">&quot;case&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">param_name</span><span class="p">:</span>
                                        <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_column_control_case&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(default: None, assumes column names are not provided)&quot;</span><span class="p">)</span>
                                    <span class="n">names_argument_added</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">param_name</span><span class="p">])</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">param_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">_column_index&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;(default: None, assumes all columns are named as expected)&quot;</span><span class="p">)</span>

        <span class="c1"># add output parameter if return type is not none</span>
        <span class="n">return_annotation</span> <span class="o">=</span> <span class="n">sig</span><span class="o">.</span><span class="n">return_annotation</span>
        <span class="k">if</span> <span class="n">return_annotation</span> <span class="o">!=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="o">.</span><span class="n">empty</span> <span class="ow">and</span> <span class="n">return_annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">return_annotation</span> <span class="o">!=</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
            <span class="c1"># Determine if it&#39;s a DataFrame or collection of DataFrames</span>
            <span class="n">is_dataframe</span> <span class="o">=</span> <span class="n">return_annotation</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
            <span class="n">is_collection</span> <span class="o">=</span> <span class="kc">False</span>
            
            <span class="k">if</span> <span class="n">get_origin</span><span class="p">(</span><span class="n">return_annotation</span><span class="p">)</span> <span class="ow">in</span> <span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                <span class="n">args</span> <span class="o">=</span> <span class="n">get_args</span><span class="p">(</span><span class="n">return_annotation</span><span class="p">)</span>
                <span class="n">is_collection</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="n">arg</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">args</span><span class="p">)</span>
            
            <span class="k">if</span> <span class="n">is_dataframe</span><span class="p">:</span>
                <span class="n">default_output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.csv&quot;</span>

                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--output&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">default_output</span><span class="p">,</span> 
                                        <span class="n">help</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Output filename (default: </span><span class="si">{</span><span class="n">default_output</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">is_collection</span><span class="p">:</span>
                <span class="n">method_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--output_prefix&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> 
                                        <span class="n">help</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Prefix for output filenames (default: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">parse_arguments</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">command</span><span class="p">:</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">print_help</span><span class="p">()</span>
        <span class="k">return</span>

    <span class="c1"># Initialize DiffMethylTools instance</span>
    <span class="n">tool</span> <span class="o">=</span> <span class="n">DiffMethylTools</span><span class="p">(</span><span class="n">pipeline</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Get the selected method</span>
    <span class="n">method</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">tool</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">command</span><span class="p">)</span>

    <span class="c1"># Prepare arguments for the method</span>
    <span class="n">sig</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="n">method</span><span class="p">)</span>
    <span class="n">method_args</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">method</span><span class="o">.</span><span class="vm">__name__</span> <span class="o">+</span> <span class="s2">&quot;.csv&quot;</span>
    <span class="k">for</span> <span class="n">param_name</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">sig</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">param_name</span> <span class="o">==</span> <span class="s2">&quot;self&quot;</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">arg_type</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">annotation</span> <span class="k">if</span> <span class="n">param</span><span class="o">.</span><span class="n">annotation</span> <span class="o">!=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="nb">str</span>

        <span class="k">if</span> <span class="n">param_name</span> <span class="o">==</span> <span class="s2">&quot;output&quot;</span> <span class="ow">or</span> <span class="n">param_name</span> <span class="o">==</span> <span class="s2">&quot;output_prefix&quot;</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="p">,</span> <span class="s2">&quot;out.csv&quot;</span> <span class="k">if</span> <span class="n">param_name</span> <span class="o">==</span> <span class="s2">&quot;output&quot;</span> <span class="k">else</span> <span class="s2">&quot;out&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        
        <span class="k">if</span> <span class="n">arg_type</span> <span class="o">==</span> <span class="n">InputProcessor</span> <span class="ow">or</span> <span class="n">arg_type</span> <span class="o">==</span> <span class="n">Optional</span><span class="p">[</span><span class="n">InputProcessor</span><span class="p">]:</span>
            <span class="n">data_indicies</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">param_name</span><span class="p">)</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">param_name</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">k</span><span class="o">.</span><span class="n">rfind</span><span class="p">(</span><span class="s2">&quot;_column_index&quot;</span><span class="p">)]</span> <span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="s2">&quot;index&quot;</span> <span class="ow">in</span> <span class="n">k</span> <span class="ow">and</span> <span class="n">param_name</span> <span class="ow">in</span> <span class="n">k</span> <span class="ow">and</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">}</span>
            
            <span class="n">data_indicies</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span> <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;coverage&quot;</span> <span class="k">else</span> <span class="s2">&quot;coverage_KEY&quot;</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">data_indicies</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
            <span class="c1"># remove trailing * if exists</span>
            <span class="c1"># data_indicies = {k[:-1] if k[-1] == &quot;*&quot; else k: v for k, v in data_indicies.items()}</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="o">+</span><span class="s2">&quot;_file&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="n">has_header</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="o">+</span><span class="s2">&quot;_has_header&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
            <span class="n">separator</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="o">+</span><span class="s2">&quot;_separator&quot;</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">separator</span> <span class="o">==</span> <span class="s2">&quot;t&quot;</span><span class="p">:</span>
                <span class="n">separator</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
                
            <span class="n">names</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="o">+</span><span class="s2">&quot;_column_names&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;ctr&quot;</span> <span class="ow">in</span> <span class="n">param_name</span><span class="p">:</span>
                <span class="n">ctr_case</span> <span class="o">=</span> <span class="s2">&quot;ctr&quot;</span>
            <span class="k">elif</span> <span class="s2">&quot;case&quot;</span> <span class="ow">in</span> <span class="n">param_name</span><span class="p">:</span>
                <span class="n">ctr_case</span> <span class="o">=</span> <span class="s2">&quot;case&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ctr_case</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="o">+</span><span class="s2">&quot;_column_control_case&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            
            <span class="nb">format</span> <span class="o">=</span> <span class="n">FormatDefinition</span><span class="p">(</span><span class="n">column_mapping</span><span class="o">=</span><span class="n">data_indicies</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">separator</span><span class="p">)</span>

            <span class="c1"># Convert CSV file paths to pandas DataFrames</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="n">file_name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">ctr_case</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">InputProcessor</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="nb">format</span><span class="p">,</span> <span class="n">has_header</span><span class="o">=</span><span class="n">has_header</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">names</span><span class="p">,</span> <span class="n">ctr_case</span><span class="o">=</span><span class="n">ctr_case</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">InputProcessor</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="nb">format</span><span class="p">,</span> <span class="n">has_header</span><span class="o">=</span><span class="n">has_header</span><span class="p">,</span> <span class="n">ctr_case</span><span class="o">=</span><span class="n">ctr_case</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">InputProcessor</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="nb">format</span><span class="p">,</span> <span class="n">has_header</span><span class="o">=</span><span class="n">has_header</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">names</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">InputProcessor</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="nb">format</span><span class="p">,</span> <span class="n">has_header</span><span class="o">=</span><span class="n">has_header</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">param_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">method_args</span><span class="p">[</span><span class="n">param_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1"># Call the method and print the result</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">method</span><span class="p">(</span><span class="o">**</span><span class="n">method_args</span><span class="p">)</span>
    <span class="c1"># TODO send to csv if dataframe output. Remember there can be multiple dataframes in output. Make default name &lt;function&gt;.csv</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">output</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="n">result</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">==</span> <span class="nb">list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">result</span><span class="p">):</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Evan Kinnear, Housemeddine Derbel, Qian Liu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>